dimod/include/dimod/adjvectordqm.h

```diff
@@ -0,0 +1,312 @@
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#ifndef DIMOD_ADJVECTORDQM_H_
+#define DIMOD_ADJVECTORDQM_H_
+
+#include <stdio.h>
+#include <algorithm>
+#include <unordered_set>
+#include <utility>
+#include <vector>
+
+#include "dimod/adjvectorbqm.h"
+#include "dimod/utils.h"
+
+namespace dimod {
+
+template <class V, class B>
+class AdjVectorDQM {
+  public:
+    using bias_type = B;
+    using variable_type = V;
+    using size_type = std::size_t;
+
+    AdjVectorBQM<variable_type, bias_type> bqm_;
+    std::vector<variable_type> case_starts_;  // len(adj_) + 1
+    std::vector<std::vector<variable_type>> adj_;
+
+    AdjVectorBQM() { case_starts_.push_back(0); }
+
+    template <class io_variable_type, class io_bias_type>
+    AdjVectorDQM(io_variable_type *case_starts, size_type num_variables, io_bias_type *linear_biases,
+                 size_type num_cases, io_variable_type *irow, io_variable_type *icol, io_bias_type *quadratic_biases,
+                 size_type num_interactions) {
+        // Set the BQM, linear biases will be added separately.
+        if (num_interactions) {
+            bqm_ = AdjVectorBQM<variable_type, bias_type>(irow, icol, num_interactions, true);
+        }
+
+        // Accounting for the cases/variables at the end without interaction.
+        while (bqm_.num_variables() < num_cases) {
+            bqm_.add_variable();
+        }
+        assert(bqm_.num_variables() == num_cases);
+
+        // Add the linear biases.
+        for (auto ci = 0; ci < num_cases; ci++) {
+            bqm_.set_linear(ci, linear_biases[ci]);
+        }
+
+        // Set the case starts.
+        case_starts_.resize(num_variables + 1);
+        for (auto v = 0; v < num_variables; v++) {
+            case_starts_[v] = case_starts[v];
+        }
+        case_starts[num_variables] = num_cases;
+
+        // Fill the adjacency list for variables.
+        std::vector<std::unordered_set<variable_type>> adjset;
+        adjset.resize(num_variables);
+        auto u = 0;
+        for (auto ci = 0, ci_end = bqm_.num_variables(); ci++) {
+            while (ci >= case_starts_[u + 1]) {
+                u++;
+            }
+            auto span = bqm_.neighborhood(ci);
+            auto v = 0;
+            while (span.first != span.second) {
+                auto cj = *(span.first).first;
+                while (cj >= case_starts_[v + 1]) {
+                    v++;
+                }
+                adjset[u].insert(v);
+                span.first++;
+            }
+        }
+
+        adj_.resize(num_variables);
+#pragma omp parallel for
+        for (auto v = 0; v < num_variables; v++) {
+            adj_[v].insert(adj_[v].begin(), adjset[v].begin(), adjset[v].end());
+            std::sort(adj_[v].begin(), adj_[v].end());
+        }
```

Comment:
Will check the self loop using a separate function since the constructor cannot return any value. The other function will. And will be called from cython.